vnext/local-cli/config/configUtils.js

```diff
@@ -0,0 +1,144 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ * @format
+ */
+// @ts-check
+
+const fs = require('fs');
+const path = require('path');
+const glob = require('glob');
+
+const xmldoc = require('xmldoc');
+
+// find VS files matching the pattern
+function findFiles(folder, filenamePattern) {
+  const files = glob.sync(path.join('**', filenamePattern), {
+    cwd: folder,
+    ignore: [
+      'node_modules/**',
+      '**/Debug/**',
+      '**/Release/**',
+      '**/Generated Files/**',
+      '**/packages/**',
+    ],
+  });
+
+  return files;
+}
+
+// finds the windows folder if present
+function findWindowsFolder(folder) {
+  const winDir = 'windows';
+  const joinedDir = path.join(folder, winDir);
+  if (fs.existsSync(joinedDir)) {
+    return joinedDir;
+  }
+
+  return null;
+}
+
+// find the visual studio solution file
+function findSolutionFile(winFolder, packageJson = {}) {
+  if (packageJson.name) {
+    const solutionName = packageJson.name;
+
+    // First look for a solution named solutionName.sln
+    const slnFile = findFiles(winFolder, solutionName + '.sln')[0];
+
+    if (slnFile) {
+      return path.join(winFolder, slnFile);
+    }
+  }
+
+  // Next look for any solution *.sln
+  const allSln = findFiles(winFolder, '*.sln');
+
+  if (allSln.length > 0) {
+    return path.join(winFolder, allSln[0]);
+  }
+
+  return null;
+}
+
+// find the visual studio project file
+function findProjectFile(winFolder, packageJson = {}) {
+  if (packageJson.name) {
+    const projectName = packageJson.name;
+
+    // First look for a project named projectName.vcxproj
+    const cppProj = findFiles(winFolder, projectName + '.vcxproj')[0];
+
+    if (cppProj) {
+      return path.join(winFolder, cppProj);
+    }
+
+    // Next look for a project named projectName.csproj
+    const csProj = findFiles(winFolder, projectName + '.csproj')[0];
+
+    if (csProj) {
+      return path.join(winFolder, csProj);
+    }
+  }
+
+  // Next look for any project *.vcxproj
+  const allCppProj = findFiles(winFolder, '*.vcxproj');
```

Comment:
I'm not a fan of the "choose the first project the filesystem gives us": More complicated apps will have multiple projects (libraries, app, support modules, etc.). Is this supposed to find the one that's the app? can we sniff which of the vcxprojs are type = Application? (is that the right thing to sniff?)

Reply:
We can technically make the heuristic as complicated as we like, but this covers the basic case: one solution, one project, named according to what the CLI produces. This backup is in-case a different name was chosen, but are still one solution, one project.

For modules/apps where this heuristic fails, they can always define the correct values in their own react-native.config file.

Reply:
the problem is that this sets up a tar pit that will take running into an issue and trying to diagnose what went wrong, whereas we know this is a problem even for an in-repo app and we can set up guardrails to protect people from setups that we can't figure out.

Reply:
Unfortunately I can't add command-line arguments into react-native config, we're just plugging into their cli. So I can make it fail if there's more than one project and we don't know which to choose, but the correct fix is for people to use react-native.config and tell us.

Reply:
Fixed: if the single correct file (solution or project) can't be idnetified, an error will be thrown and the developer will have to use the react-native.config.